File: svc-data/src/main/java/com/beancounter/marketdata/portfolio/PortfolioInputAdapter.java
package com.beancounter.marketdata.portfolio;

import com.beancounter.common.input.PortfolioInput;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.SystemUser;
import com.beancounter.common.utils.KeyGenUtils;
import com.beancounter.marketdata.currency.CurrencyService;
import java.util.ArrayList;
import java.util.Collection;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class PortfolioInputAdapter {

  private final CurrencyService currencyService;

  PortfolioInputAdapter(CurrencyService currencyService) {
    this.currencyService = currencyService;
  }

  public Collection<Portfolio> prepare(SystemUser owner, Collection<PortfolioInput> portfolios) {
    Collection<Portfolio> results = new ArrayList<>();
    for (PortfolioInput portfolio : portfolios) {
      results.add(prepare(owner, portfolio));
    }
    return results;
  }

  private Portfolio prepare(SystemUser owner, PortfolioInput portfolioInput) {
    log.debug("Creating for {}", owner.getId());
    return Portfolio.builder()
        .id(KeyGenUtils.format(UUID.randomUUID()))
        .code(portfolioInput.getCode().toUpperCase())
        .name(portfolioInput.getName())
        .currency(currencyService.getCode(portfolioInput.getCurrency()))
        .base(currencyService.getCode(portfolioInput.getBase()))
        .owner(owner)
        .build();
  }

  void fromInput(PortfolioInput data, Portfolio existing) {
    existing.setName(data.getName());
    existing.setCode(data.getCode().toUpperCase());
    existing.setCurrency(currencyService.getCode(data.getCurrency()));
    existing.setBase(currencyService.getCode(data.getBase()));
  }


}

File: svc-data/src/main/java/com/beancounter/marketdata/portfolio/PortfolioInputAdapter.kt
package com.beancounter.marketdata.portfolio

import com.beancounter.common.input.PortfolioInput
import com.beancounter.common.model.Portfolio
import com.beancounter.common.model.SystemUser
import com.beancounter.common.utils.KeyGenUtils
import com.beancounter.marketdata.currency.CurrencyService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

/**
 * Adapts PortfolioInput objects to Portfolio objects
 */
@Service
class PortfolioInputAdapter internal constructor(
    private val currencyService: CurrencyService,
    private val keyGenUtils: KeyGenUtils,
) {
    fun prepare(owner: SystemUser, portfolios: Collection<PortfolioInput>): Collection<Portfolio> {
        val results: MutableCollection<Portfolio> = ArrayList()
        for (portfolio in portfolios) {
            results.add(prepare(owner, portfolio))
        }
        return results
    }

    private fun prepare(owner: SystemUser, portfolioInput: PortfolioInput): Portfolio {
        log.debug("Creating for {}", owner.id)
        return Portfolio(
            keyGenUtils.id,
            portfolioInput.code.toUpperCase(),
            portfolioInput.name,
            currencyService.getCode(portfolioInput.currency)!!,
            currencyService.getCode(portfolioInput.base)!!,
            owner
        )
    }

    fun fromInput(data: PortfolioInput, existing: Portfolio): Portfolio {
        return Portfolio(
            existing.id,
            data.code.toUpperCase(),
            data.name,
            currencyService.getCode(data.currency)!!,
            currencyService.getCode(data.base)!!,
            existing.owner
        )
    }

    companion object {
        private val log = LoggerFactory.getLogger(PortfolioInputAdapter::class.java)
    }
}
